### ISSUE TITLE
kv: sustained write load across many ranges with high RTT replication can overload (shared) raft transport

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#111241]
High-latency links can have unexpectedly low effective throughputs due to the dynamic presented in https://github.com/cockroachdb/cockroach/issues/111238. As a result, it is possible to overload a raft transport link between two distant nodes before hitting other hardware bottlenecks first.

The purpose of this issue is to demonstrate this effect, which motivates https://github.com/cockroachdb/cockroach/issues/111238 and https://github.com/cockroachdb/cockroach/issues/111239.

### Reproduction steps:

```bash
# create a multi-region cluster
roachprod create nathan-high-rtt --nodes=4 --gce-machine-type=n2-standard-16 --clouds=gce --geo --gce-zones=us-east1-b,us-west1-b,asia-east1-a
roachprod stage  nathan-high-rtt release v23.1.10
roachprod start  nathan-high-rtt:1-3
```

This creates a cluster with the following latencies between nodes:
<img width="490" alt="Screenshot 2023-09-25 at 5 39 42 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/aefca59a-a083-432b-83b6-0a5da2bfa141">

Using the BDP math presented in https://github.com/cockroachdb/cockroach/issues/111238, we would expect that a single TCP connection between `us-east1` and `asia-east1` with a default 4MB window size could theoretically reach a throughput of `1s/187ms * 4MB = 21MB/s`. `iperf` shows 14 MB/s.

```bash
# initialize a kv table with leases in us-east
roachprod run nathan-high-rtt:1 -- './cockroach workload init kv'
roachprod sql nathan-high-rtt:1 -- -e "alter database kv configure zone using constraints = copy from parent, lease_preferences = '[[+region=us-east1]]'"

# Slowly ramp up the load using 2KB writes
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=4 --duration=1m {pgurl:1}'
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=16 --duration=1m {pgurl:1}'
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=64 --duration=1m {pgurl:1}'
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=256 --duration=1m {pgurl:1}'
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=1024 --duration=1m {pgurl:1}'
```

<details>
  <summary>This initial load test looks like the following: (click here)</summary>
<img width="983" alt="Screenshot 2023-09-25 at 4 17 37 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/e6522f2a-7495-4c6c-b567-548e7f5e60dc">
<img width="993" alt="Screenshot 2023-09-25 at 4 17 51 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/e9037180-4245-4adc-aaa3-0e86da23f039">
<img width="994" alt="Screenshot 2023-09-25 at 4 18 12 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/47730575-50dd-4ad7-8b4e-0552c58f0bd7">
<img width="981" alt="Screenshot 2023-09-25 at 4 18 27 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/6a998577-8e01-478a-bf1d-ddb82e983680">
<img width="983" alt="Screenshot 2023-09-25 at 4 18 38 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/dcec9251-f7a3-40ff-b248-8845f8d8ce75">
<img width="988" alt="Screenshot 2023-09-25 at 4 24 24 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/ebeda89b-a922-40fd-9361-5c4ab9a50c5d">
</details>

Notice that we scale up to the point where n1 (`us-east1`) is sending n3 (`asia-east1`) almost exactly 14MB. Beyond that point, we start seeing congestion in the raft transport, though it never fills up to the point of dropping messages. Recall that the raft transport is shared between ranges, so this would affect other communication between n1 and n3, not just on the `kv` range. However, the congestion never gets too bad because the quota pool kicks in and starts throttling all writes to the single `kv` range. Preventing a single slow follower from falling indefinitely behind is the purpose of the quota pool, though it's arguably a surprising effect that a minority quorum can impact availability, especially so quickly.

This already looks bad, but the quota pool is providing some protection. That's because we're only writing to a single range, so per-range limits like the quota pool and flow control in `etcd/raft` prevent sustained saturation of the raft transport. What if this exact same load was spread over 1000 ranges?
```bash
roachprod run nathan-high-rtt:1 -- './cockroach workload init --splits=1000 kv'
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=1024 --duration=1m {pgurl:1}'
```

<details>
  <summary>This test looks like the following: (click here)</summary>
<img width="989" alt="Screenshot 2023-09-25 at 6 41 16 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/ba553cc2-3e94-4da0-bb7c-50473ffffd80">
<img width="998" alt="Screenshot 2023-09-25 at 6 41 31 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/5f670c7d-fd85-4f74-8262-e175da3f1282">
<img width="985" alt="Screenshot 2023-09-25 at 6 41 42 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/f4efd0c4-94c5-4879-9387-116ab0473581">
<img width="977" alt="Screenshot 2023-09-25 at 6 41 55 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/dbfa5131-3cd2-46a1-b645-c74ccfe49d07">
<img width="982" alt="Screenshot 2023-09-25 at 6 42 04 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/b0353007-b237-4db2-8578-66a1b2fcc100">
</details>

Notice again that the n1->n3 network link only sustained about 15MB/s while the n1->n2 link sustained 20MB/s. So n3 fell further and further behind as the load ran. However, unlike the previous test, the KV table is spread across 1000 ranges, so a single range's quota pool didn't start backpressuring the workload. Instead, the load was allowed to pass through to the shared (across ranges) raft transport from n1->n3 and saturate it for the entire duration of the test. The raft transport hit its maximum size of 10k messages and then began dropping messages.

As a result, the link became effectively unusable, even for unrelated ranges (see https://github.com/cockroachdb/cockroach/issues/111239). With a queue size of 300MB and a processing rate (the network throughput) of 15MB/s, even messages from unrelated ranges that were lucky enough to accepted by the queue and not bounce off of its size limit would take over 20s to be sent out to n3. Over a sustained period of time, this could allow unrelated ranges serving a much lower rate of writes to also eventually drain their quota pool and backpressure incoming writes.

To summarize, the key ingredients to overload the raft transport between two nodes are:
- high latency
- high write throughput
- large writes
- across many ranges

Combine all four of these and the raft transport will saturate before any other hardware bottlenecks are reached.

**Comment by nvanbenschoten** [comment]
I played around with @bdarnell's suggestion from https://github.com/cockroachdb/cockroach/issues/111238#issuecomment-1734644084 to increase the TCP and gRPC (HTTP/2) window sizes. Doing so involved the following steps:

```bash
# step 1: set kernel tcp settings on all hosts
roachprod run nathan-high-rtt -- 'echo "net.ipv4.tcp_rmem = 4096 135168 33554432" | sudo tee -a /etc/sysctl.conf'
roachprod run nathan-high-rtt -- 'echo "net.ipv4.tcp_wmem = 4096 16384 33554432" | sudo tee -a /etc/sysctl.conf'
roachprod run nathan-high-rtt -- 'sudo sysctl -p'

# step 2: deploy patched crdb with https://github.com/cockroachdb/cockroach/pull/111255
roachprod put nathan-high-rtt cockroach

# step 3: start crdb with larger gRPC window size
roachprod start nathan-high-rtt:1-3 --env=COCKROACH_RPC_INITIAL_WINDOW_SIZE=33554432

# step 4: re-run workload
roachprod run nathan-high-rtt:4 -- './cockroach workload run kv --min-block-bytes=2048 --max-block-bytes=2048 --concurrency=1024 --duration=1m {pgurl:1}'
```

With those changes, the cluster was able to sustain the previous write traffic without overloading the raft scheduler.

<details>
  <summary>This looked like the following: (click here)</summary>
<img width="985" alt="Screenshot 2023-09-25 at 10 18 13 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/49585219-b764-4fb9-988d-a95abc1fa590">
<img width="983" alt="Screenshot 2023-09-25 at 10 18 34 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/2dfec710-5ec4-4bb1-bcce-1111ae4c818d">
<img width="984" alt="Screenshot 2023-09-25 at 10 18 47 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/3f3f00d9-e154-4c98-b7dc-1f45b647f94e">
<img width="987" alt="Screenshot 2023-09-25 at 10 19 05 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/1c2a2a99-dcc4-437d-a77e-243f199e6119">
</details>

Notice that the n1->n3 link was sustaining about 21MB/s.

I then increased the load further. At this point, I was saturating the CPU on the gateway/leaseholder node, so the cluster was not able to sustain the load. Still, in the brief period that it did, the n1->n3 link was transmitting at about 37MB/s.

<img width="989" alt="Screenshot 2023-09-25 at 10 27 08 PM" src="https://github.com/cockroachdb/cockroach/assets/5438456/8368490c-7642-4c4f-959c-82937402290f">

**Comment by erikgrinaker** [comment]
Thanks for the investigation and repro Nathan! We should look at making some simple, short-term improvements here.

**Comment by pav-kv** [comment]
Removing `O-support` since this issue is a repro of #111238 which we partially [mitigated](https://github.com/cockroachdb/cockroach/issues/111238#issuecomment-1977256766).

Leaving this issue open in case the mitigations end up not sufficient.
